In [165]:
import pandas as pd

In [166]:
data = {
    "Age": [25,30,35],
    "City": ['Kathmandu', 'Butwal', 'Pokhara']
}

In [167]:
df = pd.DataFrame(data, index=['Ram', 'Shyam', 'Hari'])

In [168]:
df

,Age,City
Ram,25,Kathmandu
Shyam,30,Butwal
Hari,35,Pokhara


In [169]:
df.loc['Shyam'] # row

# using .loc : Target the row by its string label

,Shyam
Age,30
City,Butwal


In [170]:
df['Age'] # column

,Age
Ram,25
Shyam,30
Hari,35


In [171]:
df.iloc[1] # Using .iloc : Target the row by its numeric position (0 is first, 1 is second).

,Shyam
Age,30
City,Butwal


In [172]:
# Slicing multiple rows

df.loc['Ram':'Shyam']

,Age,City
Ram,25,Kathmandu
Shyam,30,Butwal


In [173]:
df.iloc[1:2]

,Age,City
Shyam,30,Butwal


In [174]:
# Row and Column selection combined
# .loc : Specify row and column labels directly

df.loc['Ram','Age']

np.int64(25)

In [175]:
# Using .iloc : Use integer indices for both dimensions

df.iloc[0,1]

'Kathmandu'

In [176]:
df.iloc[1,1]

'Butwal'

In [177]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [178]:
titanic_df = pd.read_csv('/content/drive/MyDrive/dsml12/titanic-data.csv')

In [179]:
titanic_df.shape

(891, 12)

In [180]:
titanic_df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [181]:
titanic_df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [182]:
titanic_df.tail()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.00,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.00,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,NaN,Q


In [183]:
titanic_df

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


In [184]:
titanic_df.dtypes

,0
PassengerId,int64
Survived,int64
Pclass,int64
Name,object
Sex,object
Age,float64
SibSp,int64
Parch,int64
Ticket,object
Fare,float64


In [185]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [186]:
titanic_df[["Age","Fare"]].mean()

,0
Age,29.699118
Fare,32.204208


In [187]:
titanic_df[["Age","Fare"]].describe()

,Age,Fare
count,714.000000,891.000000
mean,29.699118,32.204208
std,14.526497,49.693429
min,0.420000,0.000000
25%,20.125000,7.910400
50%,28.000000,14.454200
75%,38.000000,31.000000
max,80.000000,512.329200


**Description:**
    
* Average age of passengers is approximately 29.7 years, indicating that most passengers were relatively young.
* Age distribution is fairly symmetric, as the mean (29.7) and median (28.0) are close.
* The minimum age is 0.42 years (about 5 months), and the oldest passenger is 80 years old.
* There are missing values in the Age column (714 entries vs. 891 for Fare), suggesting some age data is not available.

* For Fare, the mean is 32.20 and the median is 14.45 — a large gap indicating a right-skewed (positively skewed) distribution.
* Fare values range from 0 (free tickets) to over 512, showing a wide spread and presence of outliers or high-paying passengers.
* The high standard deviation (49.69) also supports the presence of large variability in fare prices.


In [188]:
titanic_df.value_counts("Survived")

,count
Survived,
0,549
1,342


In [189]:
titanic_df['Sex'].value_counts()

,count
Sex,
male,577
female,314


In [190]:
titanic_df.value_counts(["Sex","Survived"], ascending=True)

,,count
Sex,Survived,
female,0,81
male,1,109
female,1,233
male,0,468


In [191]:
titanic_df.value_counts(["Sex","Survived"], ascending=True ,normalize = True)

,,proportion
Sex,Survived,
female,0,0.090909
male,1,0.122334
female,1,0.261504
male,0,0.525253


In [192]:
81/891

0.09090909090909091

In [193]:
# https://github.com/datasciencedojo/datasets/blob/master/titanic.csv

In [194]:
titanic_df[titanic_df["Sex"]=="female"].value_counts("Survived",ascending=True, normalize=True)

,proportion
Survived,
0,0.257962
1,0.742038


In [195]:
titanic_df[titanic_df["Sex"]=="male"].value_counts("Survived",ascending=True, normalize=True)

,proportion
Survived,
1,0.188908
0,0.811092


In [196]:
titanic_df.value_counts("Embarked")

,count
Embarked,
S,644
C,168
Q,77


In [197]:
titanic_df[titanic_df["Embarked"]=="S"].value_counts("Sex")

,count
Sex,
male,441
female,203


In [198]:
titanic_df[titanic_df["Embarked"]=="S"].value_counts(["Survived","Sex"])

Survived  Sex   
0         male      364
1         female    140
          male       77
0         female     63
Name: count, dtype: int64

In [199]:
titanic_df[(titanic_df['Embarked'] == 'S') & (titanic_df['Sex'] == 'male')]['Survived'].value_counts()

,count
Survived,
0,364
1,77


In [200]:
df.columns = df.columns.str.strip()

In [201]:
df.columns

Index(['Age', 'City'], dtype='object')

In [202]:
import numpy as np
import pandas as pd

fare_bins = [0, 50, 100, 300, 512]
fare_labels = ["Low Fare", "Medium Fare", "High Fare", "Very High Fare"]

titanic_df["FareGroup"] = pd.cut(
    titanic_df["Fare"],
    bins=fare_bins,
    labels=fare_labels,
    include_lowest=True
)

print(titanic_df[["Fare", "FareGroup"]].head())

      Fare    FareGroup
0   7.2500     Low Fare
1  71.2833  Medium Fare
2   7.9250     Low Fare
3  53.1000  Medium Fare
4   8.0500     Low Fare


In [203]:
df["FareGroup"]

KeyError: 'FareGroup'